pkg/config/shortnames/util.go

```diff
@@ -0,0 +1,150 @@
+
+	"github.com/BurntSushi/toml"
+	"github.com/containers/storage/pkg/unshare"
+	"github.com/pkg/errors"
+)
+
+//// ALGORITHMIC UTILITY ////
+
+// linear search returning location of element in a slice (returns -1 if not found)
+func find(a []string, x string) int {
+	for i, n := range a {
+		if x == n {
+			return i
+		}
+	}
+	return len(a)
+}
+
+// removeIndex removes an element at an index in a slice
+func removeIndex(s []string, index int) []string {
+	return append(s[:index], s[index+1:]...)
+}
+
+// removeDuplicates removes duplicate values in given slice
+func removeDuplicates(s []string) []string {
+	for i := 0; i < len(s); i++ {
+		for i2 := i + 1; i2 < len(s); i2++ {
+			if s[i] == s[i2] {
+				// delete
+				s = append(s[:i2], s[i2+1:]...)
+				i2--
+			}
+		}
+	}
+	return s
+}
+
+//// FILEPATH UTILITY ////
+
+// The configuration files loaded here are used with 'user' at highest priority and 'share' least.
+
+const (
+	// _configPath is the path to the containers/shortnames.conf
+	// inside a given config directory.
+	_configPath = "containers/shortnames.conf"
+	// DefaultShortnamesConfig holds the default shortnames config path
+	DefaultShortnamesConfig = "/usr/share/" + _configPath
+	// OverrideShortnamesConfig holds the default config paths overridden by the root user
+	OverrideShortnamesConfig = "/etc/" + _configPath
+	// UserOverrideShortnamesConfig holds the containers config path overridden by the rootless user
+	UserOverrideShortnamesConfig = ".config/" + _configPath
+)
+
+// This allows the home directory to be used
+func rootlessConfigPath() (string, error) {
+	if configHome := os.Getenv("XDG_CONFIG_HOME"); configHome != "" {
+		return filepath.Join(configHome, _configPath), nil
+	}
+	home, err := unshare.HomeDir()
+	if err != nil {
+		return "", err
+	}
+
+	return filepath.Join(home, UserOverrideShortnamesConfig), nil
+}
+
+func systemConfigs() ([]string, error) {
+	configs := []string{}
+	path := os.Getenv("SHORTNAMES_CONF")
+	if path != "" {
+		if _, err := os.Stat(path); err != nil {
+			return nil, errors.Wrapf(err, "failed to stat of %s from SHORTNAMES_CONF environment variable", path)
+		}
+		return append(configs, path), nil
+	}
+	if _, err := os.Stat(DefaultShortnamesConfig); err == nil {
+		configs = append(configs, DefaultShortnamesConfig)
+	}
+	if _, err := os.Stat(OverrideShortnamesConfig); err == nil {
+		configs = append(configs, OverrideShortnamesConfig)
+	}
+	if unshare.IsRootless() {
+		path, err := rootlessConfigPath()
+		if err != nil {
+			return nil, err
+		}
+		if _, err := os.Stat(path); err == nil {
+			configs = append(configs, path)
+		}
+	}
+	return configs, nil
+}
+
+func customConfigFile() (string, error) {
+	path := os.Getenv("SHORTNAMES_CONF")
+	if path != "" {
+		return path, nil
+	}
+	if unshare.IsRootless() {
+		path, err := rootlessConfigPath()
+		if err != nil {
+			return "", err
+		}
+		return path, nil
+	}
+	return UserOverrideShortnamesConfig, nil
```

Comment:
- In this case, this file is not returned by `systemConfigs`, so the file will not be used by `Search`
- This is a path relative to home directory, hard to use by the caller. Why not return an absolute path?